### ISSUE TITLE
[docs] Document How to Use TLS + mux

### ISSUE THREAD

**Issue by elithrar** [gorilla/mux#426]
Two parts:

### Public Certs

There's a minimal example in the https://github.com/mholt/certmagic docs. We should:

* Show a complete example
* The example should enforce timeouts
* Clarify the need to have DNS set up externally

### Self-signed Certs

* How to use [mkcert](https://github.com/FiloSottile/mkcert) for dev with mux
* Generate certs
* Call `srv.ListenAndServeTLS` with said certs
* Spin up both a HTTP (:80) listener that just re-directs to HTTPS (for all routes) and a HTTPS (:443) listener concurrently.

**Comment by andrewpillar** [comment]
Hey, how much progress have you made with this?

I've got a project, [flump](https://github.com/andrewpillar/flump), that has TLS implemented with an HTTP -> HTTPS redirect handler, if TLS has been configured. You can see the logic of it in the [handler/redirect.go](https://github.com/andrewpillar/flump/blob/master/handler/redirect.go#L22) file, and the listeners getting configured in the [main.go](https://github.com/andrewpillar/flump/blob/master/main.go#L107) file too.

I would like to help with documenting some of this. Would some of the code that I have written be good enough examples for the self-signed certificates, specifically the secure redirect handler, and the concurrent listeners, provided I clean up the code a bit, and make it less specific to my use-case?

Is there also a typical structure to the documentation that should be adhered to when written, and submitted?

**Comment by elithrar** [comment]
Hey @andrewpillar - more than happy for someone to contribute this. Taking a quick look at your code, it's similar to what I was thinking from an example - but critically, I'd want the certificate provisioning to be _automatic_.

Thus:

* Use https://github.com/mholt/certmagic as a library for automatic generation of certs.
* Provide two concurrent listeners _with cancellation_ for HTTP and HTTPS
* Optional, for development: use https://github.com/FiloSottile/mkcert to set up a cert and then pass it to the `*tls.Config` of the server.

Some of the existing examples should give you guidance - e.g. https://github.com/gorilla/mux#testing-handlers:

* Heavily commented code to explain _why_
* Linked within the README at the top to support jumping down
* An explanation of what the example itself achieves

**Comment by andrewpillar** [comment]
Ok, I've forked the repository, and documented how [certmagic](https://github.com/mholt/certmagic) integration can be achieved. Added a section to the [README.md](https://github.com/andrewpillar/mux/blob/f3d0b483bebe6af3f85a88c196a7b7ce208ccd0f/README.md#automatic-https) detailing how concurrent HTTP/HTTPS listeners can be setup using [certmagic](https://github.com/mholt/certmagic) for a graceful shutdown.

Wasn't able to get around to figuring out/documenting how [mkcert](https://github.com/FiloSottile/mkcert) could be used for local development though.

Also provided additional links in the section I added to some of the Let's Encrypt documentation, if that may be of use to people.

The example I've provided is good enough to run on any server, provided you change `example.com` to the necessary domain name that points to the server. If this looks good for the public certs aspect of the documentation then I'll open up a PR.
